### ISSUE TITLE
Drop support for PHP versions less than 5.6.

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#3992]
Sibling issue to #214 (I thought we already had one for this, but I could not find anything).

None of the tools that I use to set up local environments (ddev, lando etc) support php versions below 5.6, which makes it hard for me to test things when there are failures in my PRs for php53, but not for php70.

Here's a few stats from around the net, to help us decide...

https://blog.packagist.com/php-versions-stats-2019-1-edition

![image](https://user-images.githubusercontent.com/2423362/63644020-902b1b80-c6e7-11e9-8e7c-4ad308b20045.png)

>A few observations: PHP 7.3 is growing almost as well as 7.2 last year. With PHP 5.6 and 7.0 reaching end of life at the end of 2018, the number of people using maintained PHP versions dropped to 80% (from 84%), which is really not as bad as I expected as 31% of people had to upgrade. PHP 5 usage across all versions is now around 10%.

https://w3techs.com/technologies/details/pl-php/all/all

<img width="363" alt="Screen Shot 2019-08-25 at 3 25 09 am" src="https://user-images.githubusercontent.com/2423362/63644081-42171780-c6e9-11e9-8bcc-f166c11b1ed1.png">

https://w3techs.com/technologies/details/pl-php/5/all

<img width="396" alt="Screen Shot 2019-08-25 at 3 34 27 am" src="https://user-images.githubusercontent.com/2423362/63644082-47746200-c6e9-11e9-8382-3ed7c98a4b53.png">

https://wordpress.org/about/stats

<img width="459" alt="Screen Shot 2019-08-25 at 3 29 03 am" src="https://user-images.githubusercontent.com/2423362/63644084-4d6a4300-c6e9-11e9-97e9-5eb19dc55b6f.png">

---

- [x] Initial warning PR: https://github.com/backdrop/backdrop/pull/3881
- [ ] Actual increase of minimum version in bootstrap.inc: https://github.com/backdrop/backdrop/pull/3922

**Comment by klonos** [comment]
...it's also worth mentioning that both php 5.6 and php 7.0 have been EoL for a while, and php 7.1 will be EoL by the end of the year.

**Comment by indigoxela** [comment]
> both php 5.6 and php 7.0 have been EoL for a while

While that's correct regarding _php.net_, it's only half of the truth. ;)

Some Linux distributions still maintain older php versions and take care of security backports (think of CentOS 6 or Debian LTS).

> ...makes it hard for me to test things when there are failures in my PRs for php53...

So true! Unless using CentOS 6 (which has php53 as default), it's hard to get PHP older than 5.6 but still maintained.

**Comment by klonos** [comment]
Now that #285 is in core, we should be able to make a better-informed decision (once we have gathered enough metrics that is).

**Comment by klonos** [comment]
Stats update:

- https://blog.packagist.com/php-versions-stats-2021-1-edition
  ![image](https://user-images.githubusercontent.com/2423362/133891407-ab35121b-737e-4a8b-8b49-2078e55d01c9.png)
- https://w3techs.com/technologies/details/pl-php
  ![image](https://user-images.githubusercontent.com/2423362/133891419-b2f17eff-daf1-43bc-ba00-d5690f05e55f.png)
- https://w3techs.com/technologies/details/pl-php/5
![image](https://user-images.githubusercontent.com/2423362/133891426-69c3d2a1-6a9a-4b2f-abf4-560581ec5fdf.png)
- https://wordpress.org/about/stats
  - 5.6 -> 8.3%
  - 5.5 -> 0.8%
  - 5.4 -> 1.1%
  - 5.3 -> 1.2%
  - 5.2 -> 0.5%

**Comment by klonos** [comment]
...I'd say that it's safe to deduct that the percentage of the internet still stuck with php5.5 and earlier is less than 15%.

**Comment by philsward** [comment]
Take into consideration people who have the ability to upgrade, but don't know they can or don't think they can, so they stay on older versions.

I'm all for stepping up the requirements. For anyone migrating from D7 over the next few years, I think even making the requirement set to 7, won't affect many people overall. If the requirement for Backdrop is php7, and someone needs to migrate, the migration is the perfect time to update the platform anyway.

I don't know if there is a way for telemetry to "ask" for all supported versions on the system or not, but if that is a possibility, it would be good to collect the used version of PHP as well as the available versions to the system to get a better idea of how many people are on an older version but have direct access to a newer version. CPanel and Plesk for example allow multiple versions on the same system.

Long story short, I'm all for tightening up the minimum requirements and would prefer to see it tightened up even further to v7. We are finally in a day-and-age where the minimum version isn't the target, but rather the latest working version that won't break things is the desired target on account of the speed boosts each newer version provides.

Also, if we can add telemetry to the Backdrop Upgrade Status module for D7, we would have a better idea of the hurdles this discussion might have for D7 migrations, helping even further to make an informed decision.

**Comment by klonos** [comment]
> Also, if we can add telemetry to the Backdrop Upgrade Status module for D7, we would have a better idea of the hurdles this discussion might have for D7 migrations, helping even further to make an informed decision.

That's a good idea/point 👍🏼

**Comment by quicksketch** [comment]
I couldn't find my old issue where I laid out our PHP version policy, but the previous policy had been to support the oldest still-supported version of PHP provided by a Linux distribution. Here's a new table of default PHP versions by distribution:

| Distro      | Release Date | EOL Date   | PHP Version |
| ----------- | ------------ | ---------- | ----------- |
| CentOS 6    | 2011-07-10   | 2020-11-30 | 5.3         |
| CentOS 7    | 2014-07-07   | 2024-06-30 | 5.4         |
| CentOS 8    | 2019-09-24   | 2021-12-31* | 7.3         |
| RHEL 6      | 2010-11-10   | 2020-11-30 | 5.3         |
| RHEL 7      | 2014-06-10   | 2021-08-30 | 5.4         |
| RHEL 8      | 2019-05-07   | 2029       | 7.3         |
| Debian 8    | 2015-04-25   | 2020-06-30 | 5.6         |
| Debian 9    | 2017-06-17   | 2022-06-30 | 7.0         |
| Debian 10   | 2019-07-06   | 2024-06    | 7.2         |
| Debian 11   | 2021-08-14   | 2026-06    | 7.4         |


*CentOS 8 were discontinued by RedHat, and the planned 10-year support cycle will be cut-off at the end of 2021 instead.

Ubuntu more or less follows the same support cycle as Debian, so I didn't include it here.

Due to the extremely long cycles of CentOS, that previously had been our benchmark which version of PHP we support. Since CentOS 6 went EOL last year, we should have at least bumped up to PHP 5.4 as the minimum version.

If we were to stick with the current policy, CentOS 7 is the limiting OS with PHP 5.4 as the default. However, the popularity of CentOS has waned quite a bit in recent years and is now discontinued. I think the compromise of bumping up to 5.6 would be a good one.

**Comment by herbdool** [comment]
Reading through the comments it seems that nobody is opposed to the main suggestion: drop support for PHP below 5.6.

There are other good discussions but perhaps new issues can be created for those.

**Comment by quicksketch** [comment]
I'm agree with this approach. Let's start with dropping support below PHP 5.6.

I'd love if the PMC can confirm this is acceptable, since it's a direct question as to how far we pursue our backwards-compatibility principle.

We'll need a longer runway than 1.21.0 though. I would prefer to commit this as an early change in the 1.22.0 cycle (meaning we'd drop support May 15, 2022)

**Comment by BWPanda** [comment]
Seconded ^

**Comment by stpaultim** [comment]
I've opened an issue in the PMC issue queue and asked for a vote to restrict support for PHP versions 5.6 or greater beginning May 15th, 2022.

**Comment by stpaultim** [comment]
The PMC approved this and we posted an announcement. 
https://backdropcms.org/news/backdrop-cms-drops-support-for-php-versions-prior-to-56

**Comment by stpaultim** [comment]
@klonos or @quicksketch - Did we decide to implement some kind of warning or error message on the status page for sites that are running very old versions of PHP. 

Is there an issue for that yet? I assume we need to create that issue as a follow-up to this change.

**Comment by domaingood** [comment]
@stpaultim  warning  message on "Outdated PHP Version" , not error message.
Also Recommend Supported PHP Versions from https://www.php.net/supported-versions.php. Always

**Comment by klonos** [comment]
> Is there an issue for that yet?

This issue here is labeled as `pr - ready to be committed`, but there's not linked or other PR.

We could use #3490, for which I have a PR (will need to be rebased and tweaked accordingly).

**Comment by herbdool** [comment]
Since @klonos has linked to an issue that means we can close this one. I don't think there's anything else you to do?

**Comment by klonos** [comment]
There's a couple fo things remaining to be done:
- update https://docs.backdropcms.org/documentation/system-requirements (not a core issue)
- update the `define('BACKDROP_MINIMUM_PHP', '5.3.2');` line in https://github.com/backdrop/backdrop/blob/1.x/core/includes/bootstrap.inc#L20

**Comment by jenlampton** [comment]
As per today's meeting, this issue needs a PR to add the message into the status report.

**Comment by quicksketch** [comment]
I updated our official requirements page with the new PHP version: https://docs.backdropcms.org/documentation/system-requirements

I filed a PR at https://github.com/backdrop/backdrop/pull/3881 that shows the PHP version warning if PHP is lower than 5.6.0 (should the bugfix version be higher here, like 5.6.10 or something?)

Here's the status report when running PHP 5.3:

![Screenshot from 2021-12-30 19-35-09](https://user-images.githubusercontent.com/100206/147801741-06b14d14-8916-4ce2-b46e-814ff1d8cb64.png)

I discovered that Backdrop will refuse to install or run updates if we have an error in the status report, so if the site is not in "runtime" (normal) use, the error is downgraded to a warning when visiting install.php or update.php:

![Screenshot from 2021-12-30 19-24-26](https://user-images.githubusercontent.com/100206/147802186-e15e3c84-c0d7-442a-9325-473f7d7b96a2.png)

Warnings apparently can be skipped by clicking the "try again" link at the bottom of the page, which will allow updates to be run.

![Screenshot from 2021-12-30 19-24-59](https://user-images.githubusercontent.com/100206/147802198-b3374d9e-767c-4bce-9c19-e35405f4c892.png)

Perhaps we could improve this by changing the wording here to "[try again](#) or [proceed anyway](#)" if only warnings are present?

**Comment by jenlampton** [comment]
There are some weird test failures on the PR, `Class 'Entity' not found`?  Can you push your branch again @quicksketch  to restart the tests?

**Comment by quicksketch** [comment]
I'm working on fixing the test failures on PHP 5.3. Now that this warning is displayed, the Upgrade tests are failing (since there's now an extra step in clicking "try again" for deprecated PHP versions).

**Comment by quicksketch** [comment]
Okay it looks like there are a slew of Upgrade tests that fail if we throw a warning on the `update.php` page, so instead of building in a dozen exceptions, I downgraded the "warning" to a "notice" when running updates. The PHP Version is still an error on new installs and at runtime (the status report).

**Comment by quicksketch** [comment]
All tests are passing now with the new approach of only throwing a runtime error on the status report.

**Comment by herbdool** [comment]
I've tested locally and reviewed code. Looks good to me!

**Comment by quicksketch** [comment]
Super, thanks @herbdool! Keeping this open and moving to 1.22.0.

**Comment by quicksketch** [comment]
PR for bumping the minimum to PHP 5.6.0: https://github.com/backdrop/backdrop/pull/3922

I couldn't find a reasonable specific minimum bugfix version (like 5.6.4 versus 5.6.20, etc) there are number of different versions included by default in various Linux distros and Mac OS, so just 5.6.0 and higher seems like the best option.

**Comment by quicksketch** [comment]
I searched for "5.3", "5.4", and "5.5" in the code base and started removing version-specific work-arounds and the PR quickly became _much_ larger. Now the PR includes all version-specific fixes I could find, but we might want to separate out the removing/deprecating of code into a separate issue rather than include it all here.

**Comment by quicksketch** [comment]
I reworked the PR to remove the most drastic change: deprecation of `backdrop_basename()`. It turns out this function has substantial differences from PHP's `basename()` which provides a good reason to keep it even in new PHP versions. This makes for a lot fewer changes and hopefully solves most test failures, which helpfully exposed the differences.

**Comment by klonos** [comment]
Thank you @quicksketch 👍🏼  ...not php7.4 and 8.1 tests pass - only php5.6 fails.

**Comment by quicksketch** [comment]
The PHP 5.6 test runs failing has got me stumped. I filed another PR at https://github.com/backdrop/backdrop/pull/3936 that makes _no_ changes except running the tests on PHP 5.6, and they are all failing. I tried running on PHP 5.5 however, and that ran perfectly fine. I think there may be some inconsistency with the `setup-php` action we're using, or a default configuration of PHP 5.6 is not compatible with our tests. It seems to be something about file system permissions.

**Comment by indigoxela** [comment]
> It seems to be something about file system permissions.

Strange, no files can get written - not in the files directory, not even temporary. Maybe the GHA setup (php.ini?) needs tweaks. Will have a look ASAP.

**Comment by indigoxela** [comment]
I can fix the file permission problem for php: [PR](https://github.com/backdrop/backdrop/pull/3940).

But there's more going on which I can't fix.

In the first test fraction I get a bunch of "Class not found" errors - no clue why.
https://github.com/backdrop/backdrop/runs/5048318802?check_suite_focus=true

In the "Setup Webserver" section of that run I tried to debug php a bit more - a list of loaded modules (php -m) and the full php.ini file. Still no idea, what might be the cause.

**Comment by indigoxela** [comment]
That's a general problem with our tests! Even if "tests are green", there are actually all these "class not found" errors (mostly with tests in core/modules/system/tests/system.test it seems). Also on other php versions in our GHA, also on my local dev - with any php version. It's only that php 5.6 now chokes on these.

**Comment by klonos** [comment]
Hmm. Not good 🤔

**Comment by indigoxela** [comment]
I opened a separate issue for that: #5490

Hm... should I include the fix for GHA (php ini setup) in that other PR?

**Comment by klonos** [comment]
> should I include the fix for GHA (php fpm setup) in that other PR?

I just tested and RTBC'ed the PR for #5490, and I plan to bring it up during the dev meeting later today, so I expect it to be merged soon. Then all you'll need to do is rebase the GHA PR 😉

**Comment by quicksketch** [comment]
Thank you SO MUCH @indigoxela! I couldn't figure out why this was GHA-specific, but that fix in `.github/misc/setup-php-fpm.sh` explains it! I incorporated your fix into my PR at https://github.com/backdrop/backdrop/pull/3922.

I also had to restore the GD imagerotate() work-around, as PHP 5.6 on GHA gets a different result than my local and newer versions of PHP 7. All tests should now be passing in the PR!

**Comment by indigoxela** [comment]
Regarding my questions about the possible remove of the error control operator: it's probably OK to leave them in place. They don't break anything - removing on the other hand could possibly (although it's unlikely).

There's actually nothing to actively test here. RTBC! :+1:

**Comment by BWPanda** [comment]
Based on @indigoxela's RTBC label, and @quicksketch's comments in today's meeting re. just merging this in. I've gone ahead and done so.

Thanks to @quicksketch, @klonos, @indigoxela & @herbdool  for the PR and reviews. Thanks also to everyone else for contributing and giving feedback. I've merged https://github.com/backdrop/backdrop/pull/3922 into 1.x.

**Comment by quicksketch** [comment]
Thanks @indigoxela and @BWPanda!
